Add ScopedUpdater components to improve re-render performance

cottom opened the pull request:
## Summary

Previously we have meet some cases that interaction in some page is not in time which makes users feel restlessness. The cause is that goji-runtime always update UI in the page scope. While as the page DOM amount increased, time cost  for update every time render can increase exponentially as well. 

This PR is to add a component to limit the update scope manually by developer.

## Usage

we provide `ScopeUpdater` as a react component and it has a corresponding component `scope-updater` in `miniapp-runtime. Every UI update under this component will be sent by `scope-update`  as a host.

```tsx
<Parent>
  <Bala/>
  <ScopeUpdater>
    <Timmer />
  </ScopeUpdater>
  <Balabala/>
<Parent/>
```

## implementation

1. [webpack-plugin] add `scope-updater` component in `goji-runtime`. Add `isHostWrapped` props to describe wrapper component who need to handle events and styles by itself. 
2. [core] add `sendUpdate` to process the `diff`, pickup the host `scopeUpdate` component by update path and update the diff just in the scope.

## Benchmark

#### environment
MacBook Pro (16-inch, 2019)
processor 2.4 GHz 8-Core Intel Core i9
Memory 32 GB 2667 MHz DDR4



#### test case

```tsx
export const BenchmarkControl: React.FC<{ isTreatment: boolean; nestCount: number }> = ({ isTreatment, nestCount }) => {
  const [num, setNum] = useState(0);

  const child = isTreatments ? (
    <ScopeUpdater>
      <View>{num}</View>
    </ScopeUpdater>
  ) : (
    <View>{num}</View>
  );

  const loop = (count: number) => {
    return (
      <View className={styles.container}>
        <View>
          <View>{count === 0 ? child : loop(count - 1)}</View>
          <View />
        </View>
      </View>
    );
  };

  return (
    <>
      {loop(nestCount)}
      <Button onTap={() => setNum(i => i + 1)}>++</Button>
    </>
  );
};

```

![image](https://user-images.githubusercontent.com/11937539/113132134-a187c280-9250-11eb-883a-0cd8a0ca82dc.png)

#### prod case (P2 filter)

![p2-filter](https://user-images.githubusercontent.com/11937539/113157462-7bbce680-926d-11eb-94ea-827b4803b7d4.gif)

malash reviewed (changes requested):
Thanks @cottom for your PR. I think it's a good creation to solve update performance in GojiJS. But, there are several problems we have to resolve.

1. Could you please rename the component to `ScopedUpdater` ?
2. Use `sendUpdateData` to transform JSON data is not a good idea. I prefer to modify `pure` function to support this feature.
3. Advice to use `Map<ComponentInstance, Updates>` for `scopedUpdates`
4. Add test cases for this new feature.
5. [useRef](https://goji.js.org/docs/en/refs) may not work inside `ScopedUpdater`.
6. Did you try this component on all platforms ?

malash commented on packages/webpack-plugin/templates/components/scroll-view.js.ejs:
```diff
@@ -8,7 +8,7 @@ Component({
   },
   properties: {
-    nodes: {
+    c: {
```
Why this change ?

cottom replied:
unified all `nodes` properties

malash commented on packages/core/src/reconciler/instance.ts:
```diff
@@ -48,6 +48,10 @@ export abstract class BaseInstance {
     this.parent = parent;
   }
 
+  public getParent() {
```
Useless function

cottom commented:
> Thanks @cottom for your PR. I think it's a good creation to solve update performance in GojiJS. But, there are several problems we have to resolve.
> 
> 1. Could you please rename the component to `ScopedUpdater` ?
> 2. Use `sendUpdateData` to transform JSON data is not a good idea. I prefer to modify `pure` function to support this feature.
> 3. Advice to use `Map<ComponentInstance, Updates>` for `scopedUpdates`
> 4. Add test cases for this new feature.
> 5. [useRef](https://goji.js.org/docs/en/refs) may not work inside `ScopedUpdater`.
> 6. Did you try this component on all platforms ?

1. NIT, updated.
2. change pure is not easy to meet the demand of the feature and also break lots of test case. oervall pure is not a good practice here. Just leave it for later refractor.
3. NIT, updated.
4. updated.
5. updated.
6. Yes.

malash commented on packages/webpack-plugin/templates/leaf-components.wxml.ejs:
```diff
@@ -2,16 +2,18 @@
   <block wx:elif="{{<% if (component.sid === undefined) {%>type === '<%= component.name %>'<%} else {%>sid === <%= component.sid %><%}%>}}">
     <<%= component.isWrapped ? 'goji-' : ''%><%= component.name %>
-      <% if (component.isWrapped) { %>
+      <% if (component.isWrapped && !component.isSelfHost) { %>
```
isHostWrapped

malash commented on packages/core/src/components/scopedUpdater.tsx:
```diff
@@ -0,0 +1,35 @@
+  }
+
+  return createElement(Fragment, {
+    ref,
```
No need to forward ref here

malash commented on packages/core/src/components/scopedUpdater.tsx:
```diff
@@ -0,0 +1,35 @@
+}>;
+
+const SelfScopedUpdater = (
+  props: ScopedUpdaterProps,
```
Use `unsafe_className` and `unsafe_style`

malash commented on packages/webpack-plugin/src/constants/components.ts:
```diff
@@ -604,8 +605,21 @@ export const getBuiltInComponents = (target: GojiTarget): ComponentDesc[] =>
   );
 
+export const getExtendComponents = (target: GojiTarget): ComponentDesc[] => {
+  return [
```
`['wechat', 'qq'].includes(target) ? [{ xxx }] : []`

malash commented on packages/webpack-plugin/templates/leaf-components.wxml.ejs:
```diff
@@ -2,16 +2,18 @@
         class="{{props.className}}"
         id="{{props.id}}"
         style="{{props.style || ''}}"
+        c="{{c}}"
```
Remove this line
